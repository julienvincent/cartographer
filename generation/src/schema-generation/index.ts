export * as litematica from './litematica';
export * as map from './map';
