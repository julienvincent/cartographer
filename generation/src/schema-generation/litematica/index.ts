import * as litematica from '@cartographer/litematica';
import { MCBlockSpace } from '../../block-generation';

/**
 * Converts BlockSpace data into a litematica schema. The resulting schema data will need to be serialized
 * using the 'nbt' module before exporting
 */
export const generateLitematicaSchema = (block_space: MCBlockSpace) => {
  const blocks = block_space.reduce((blocks: litematica.BlockPosition[], columns, x) => {
    return columns.reduce((blocks, rows, z) => {
      return rows.reduce((blocks, block) => {
        blocks.push({
          id: block.id,
          properties: block.properties,
          x,
          y: block.y_offset,
          z
        });
        return blocks;
      }, blocks);
    }, blocks);
  }, []);

  return litematica.generateSchematicNBT({
    blocks,
    name: 'Map',
    author: 'Cartographer',
    description: 'MapArt schematic generated by Cartographer'
  });
};
