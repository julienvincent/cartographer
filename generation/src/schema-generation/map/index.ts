import { MCBlockSpace } from '../../block-generation';

const getBlockSpaceHeight = (block_space: MCBlockSpace) => {
  return block_space.reduce((height, columns) => {
    return columns.reduce((height, rows) => {
      return rows.reduce((height, pillar) => {
        if (pillar.height + 1 > height) {
          return pillar.height + 1;
        }
        return height;
      }, height);
    }, height);
  }, 0);
};

export const asNbtObject = (space: MCBlockSpace): object => {
  const width = space.length;
  const length = space[0].length;
  const height = getBlockSpaceHeight(space);

  const palette: string[] = ['air'];

  const getPaletteId = (sid: string): number => {
    let id: number = palette.indexOf(sid);
    if (id === -1) {
      id = palette.length;
      palette.push(sid);
    }
    return id;
  };

  const getSid = (x: number, y: number, z: number): string => {
    const pillar = space[x][z];

    for (let b of pillar) {
      if (b.height === y) return b.block_id;
    }

    return 'air';
  };

  const blocks: object[] = (() => {
    const arr: object[] = [];

    for (let y = 0; y < width; y++) {
      for (let z = 0; z < length; z++) {
        for (let x = 0; x < width; x++) {
          arr.push({
            pos: {
              type: 'list',
              value: {
                type: 'int',
                value: [x, y, z]
              }
            },
            state: {
              type: 'int',
              value: getPaletteId(getSid(x, y, z))
            }
          });
        }
      }
    }

    return arr;
  })();

  return {
    name: '',
    value: {
      size: {
        type: 'list',
        value: {
          type: 'int',
          value: [width, height, length]
        }
      },
      entities: {
        type: 'list',
        value: {
          type: 'end',
          value: []
        }
      },
      blocks: {
        type: 'list',
        value: {
          type: 'compound',
          value: blocks
        }
      },
      palette: {
        type: 'list',
        value: {
          type: 'compound',
          value: palette.map((s) => {
            return {
              Name: {
                type: 'string',
                value: 'minecraft:' + s
              }
            };
          })
        }
      },
      DataVersion: {
        type: 'int',
        value: 2584
      }
    }
  };
};
