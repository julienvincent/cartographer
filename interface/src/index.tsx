import * as ReactDOM from 'react-dom';
import * as React from 'react';
import * as Root from './root';

// import 'nes.css/css/nes.min.css';
import './index.css';

ReactDOM.render(
  <React.StrictMode>
    <Root.Component />
  </React.StrictMode>,
  document.getElementById('root')
);
