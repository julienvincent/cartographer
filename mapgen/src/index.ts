export * as schema_generation from './schema-generation';
export * as block_generation from './block-generation';
export * as serialization from './serialization';
