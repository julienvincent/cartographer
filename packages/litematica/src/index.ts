export * as litematica_bit_array from './litematica-bit-array';
export * as long from './long';
export * from './generator';
