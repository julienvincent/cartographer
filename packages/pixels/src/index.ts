export * as transformers from './transformers';
export * as conversion from './conversion';
export * from './definitions';
