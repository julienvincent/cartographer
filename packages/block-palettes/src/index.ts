import { BlockPalette } from '@cartographer/pixels';

import palette_1_20 from './palettes/1.20.json';
import palette_1_19 from './palettes/1.19.json';
import palette_1_18 from './palettes/1.18.json';

export const palettes = {
  '1.20': palette_1_20 as unknown as BlockPalette,
  '1.19': palette_1_19 as unknown as BlockPalette,
  '1.18': palette_1_18 as unknown as BlockPalette
};
