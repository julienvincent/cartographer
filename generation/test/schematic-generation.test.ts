import { describe, it, expect } from 'vitest';

import * as schematic_generation from '../src/schema-generation';

describe('schematic-generation', () => {
  it('should generate a schematic file', () => {
    const data = schematic_generation.litematica.generateSchematicNBT([
      [
        [
          {
            id: 'minecraft:stone',
            y_offset: 0
          },
          {
            id: 'minecraft:birch_log',
            properties: {
              axis: 'y'
            },
            y_offset: 1
          }
        ],
        [
          {
            id: 'minecraft:white_wool',
            y_offset: 0
          }
        ]
      ],
      [
        [],
        [
          {
            id: 'minecraft:birch_log',
            properties: {
              axis: 'y'
            },
            y_offset: 0
          },
          {
            id: 'minecraft:birch_log',
            properties: {
              axis: 'x'
            },
            y_offset: 1
          }
        ]
      ]
    ]);

    expect(data).toMatchSnapshot();
  });
});
