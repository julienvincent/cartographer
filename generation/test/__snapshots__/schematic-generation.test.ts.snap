// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schematic-generation it should generate a schematic file 1`] = `
Object {
  "name": "",
  "value": Object {
    "Metadata": Object {
      "type": "compound",
      "value": Object {
        "Author": Object {
          "type": "string",
          "value": "cartographer",
        },
        "Description": Object {
          "type": "string",
          "value": "",
        },
        "EnclosingSize": Object {
          "type": "compound",
          "value": Object {
            "x": Object {
              "type": "int",
              "value": 2,
            },
            "y": Object {
              "type": "int",
              "value": 2,
            },
            "z": Object {
              "type": "int",
              "value": 2,
            },
          },
        },
        "Name": Object {
          "type": "string",
          "value": "map",
        },
        "RegionCount": Object {
          "type": "int",
          "value": 1,
        },
        "TimeCreated": Object {
          "type": "long",
          "value": Array [
            375,
            2023476248,
          ],
        },
        "TimeModified": Object {
          "type": "long",
          "value": Array [
            375,
            2023476248,
          ],
        },
        "TotalBlocks": Object {
          "type": "int",
          "value": 4,
        },
        "TotalVolume": Object {
          "type": "int",
          "value": 8,
        },
      },
    },
    "MinecraftDataVersion": Object {
      "type": "int",
      "value": 2584,
    },
    "Regions": Object {
      "type": "compound",
      "value": Object {
        "map": Object {
          "type": "compound",
          "value": Object {
            "BlockStatePalette": Object {
              "type": "list",
              "value": Object {
                "type": "compound",
                "value": Array [
                  Object {
                    "Name": Object {
                      "type": "string",
                      "value": "minecraft:air",
                    },
                  },
                  Object {
                    "Name": Object {
                      "type": "string",
                      "value": "minecraft:stone",
                    },
                  },
                  Object {
                    "Name": Object {
                      "type": "string",
                      "value": "minecraft:oak_planks",
                    },
                  },
                  Object {
                    "Name": Object {
                      "type": "string",
                      "value": "minecraft:white_wool",
                    },
                  },
                  Object {
                    "Name": Object {
                      "type": "string",
                      "value": "minecraft:cobblestone",
                    },
                  },
                ],
              },
            },
            "BlockStates": Object {
              "type": "longArray",
              "value": Array [
                Array [
                  1,
                  10433,
                ],
                Array [
                  0,
                  0,
                ],
              ],
            },
            "Entities": Object {
              "type": "list",
              "value": Object {
                "type": "end",
                "value": Array [],
              },
            },
            "PendingBlockTicks": Object {
              "type": "list",
              "value": Object {
                "type": "end",
                "value": Array [],
              },
            },
            "PendingFluidTicks": Object {
              "type": "list",
              "value": Object {
                "type": "end",
                "value": Array [],
              },
            },
            "Position": Object {
              "type": "compound",
              "value": Object {
                "x": Object {
                  "type": "int",
                  "value": 1,
                },
                "y": Object {
                  "type": "int",
                  "value": 0,
                },
                "z": Object {
                  "type": "int",
                  "value": 1,
                },
              },
            },
            "Size": Object {
              "type": "compound",
              "value": Object {
                "x": Object {
                  "type": "int",
                  "value": 2,
                },
                "y": Object {
                  "type": "int",
                  "value": 2,
                },
                "z": Object {
                  "type": "int",
                  "value": 2,
                },
              },
            },
            "TileEntities": Object {
              "type": "list",
              "value": Object {
                "type": "end",
                "value": Array [],
              },
            },
          },
        },
      },
    },
    "Version": Object {
      "type": "int",
      "value": 5,
    },
  },
}
`;
